### PR TITLE
Development Branch

### PR DESCRIPTION
**Pull request by mmstick**
Branch for continuous testing

### CODE REVIEW

**Comment by brs17** [comment]
Did not complete regression testing for 2020-05-06. This is what I have found thus far regardless.

I have yet to be able to replicate the issues the following commits attempt to solve
https://github.com/pop-os/shell/pull/300/commits/255c01d11c6f4f327e93137b310201ee42351635
https://github.com/pop-os/shell/pull/300/commits/0cda51419689c2bf88c0588e4e1aa9cff09fcf7c
https://github.com/pop-os/shell/pull/300/commits/4e60d4ca647141e308ea313074e3d0adf097f90b

- Firefox
  - With Firefox and Firefox nightly tiled, toggling showing the title bar swaps the two windows. This works with either Firefox browser
  - Hiding the title bar does hide a second close button and seems to behave(other than the above mentioned issue) in non-English languages for both Firefox and Firefox nightly


- With a few windows tiled on an external display, changing the resolution of that display moves all of the windows to the primary display
  - This happens whether increasing or decreasing the resolution


- With launching snaps from the launcher, I tried launching keepassxc and inkscape. Both programs were missing their icons in the launcher. They did however launch successfully
  - Launching programs from /usr/local seems to work as expected


- I see that a number of logs have been removed (as expected). 
  - logs still appear for changing window focus
  - logs still show for moving windows across forks
  - This is a good first step, but as I had mentioned elsewhere, I still think logging the title of each window that gains focus is too much


- Gedit does not open properly. With nothing else is open, launching gedit shows the active hint on the primary display but the window is not there. It looks like the window is one workspace below but trying to change workspaces or selecting the window just leads to oddities where I am still unable to access gedit

File: src/extension.ts [hunk 1]
```diff
@@ -1326,13 +1346,39 @@ function find_unused_workspace(): [number, any] {
...
+            }
+
+            // Merge theme update with pop shell styling
+            existing_theme.load_stylesheet(STYLESHEETS[pop_stylesheet]);
+
+            // Perform theme update
+            theme_context.set_theme(existing_theme);
+        } else {
+            // User does not have a theme loaded, so use pop styling + default
+            setThemeStylesheet();
```
Comment (jambonmcyeah) [inline src/extension.ts:1377]: I think you meant to write `setThemeStylesheet(pop_stylesheet_path)`

**Comment by brs17** [comment]
Finished testing from yesterday (2020-05-07). The following feedback as well as feedback listed here: https://github.com/pop-os/shell/pull/300#issuecomment-624973902 is what I have.

NOTE: This only includes commits from yesterday.

## With tiling enabled

### Tiling

- [x] Super direction keys changes focus in the correct direction
- [x] Windows moved with the keyboard tile into place
- [x] Windows moved with the mouse tile into place
- [x] Windows swap with the keyboard (test with different size windows)
- [x] Windows can be resized with the keyboard (Test resizing four windows above, below, right, and left to ensure shortcut consistency)
- [x] Windows can be resized with the mouse
- [x] Minimizing a window detaches it from the tree and re-tiles remaining windows
- [FAIL] Unminimizing a window re-tiles it into its previous location
- [x] Maximizing removes the active hint and covers tiled windows
- [x] Umaximizing adds active hint and re-tiles into place
- [x] Full-screening removes the hint and full-screens on one display (Not required, currently has a bug)
- [x] Unfull-screening adds active hint and re-tiles into place (Not required, currently has a bug)
- [x] Maximizing a YouTube video fills the screen and unmaximizing retiles the browser in place
- [x] VIM shortcuts work as direction keys
- [?] `Super` `O` changes window orientation
- [x] `Super` `G` floats and then re-tiles a window
- [x] `Super` `Q` Closes a window
- [x] `Super` `M` Maximizes and un-maximizes a window
- [x] Turn off auto-tiling. New windows launch floating.
- [x] Turn on auto-tiling. Windows automatically tile.
- [ ] Disabling and enabling auto-tiling correctly handles minimized, maximized, fullscreen, floating, and non-floating windows
      - minimizing a window and then toggling tiling off and then on again, show the window (although I don't think actually takes it out of minimize
      - maximized window is un maximized when autotiling is toggled off then on again

### Workspaces

- [ ] Windows can be moved to another workspace with the keyboard
      - when workspace is moved to new workspace, moving it back gets weird things to happen (see video)
[workspaces.zip](https://github.com/pop-os/shell/files/4594751/workspaces.zip)
- [x] Windows can be moved to another workspace with the mouse
NOTE: With issues like seen in the video above, testing workspaces became very difficult as I was struggling to access open applications. After half a dozen restarts of both gnome and the system I decided to mark the follow tests as failures as I was spending way too much time trying to get workspaces working

- [ ] Windows can be moved to workspaces between existing workspaces
- [ ] Moving windows to another workspace re-tiled the previous and new workspace
- [ ] Active hint is present on the new workspace and once the window is returned to its previous workspace
- [ ] Floating windows move across workspaces
- [ ] Remove windows from the 2nd worspace in a 3 workspace setup. The 3rd workspace becomes the 2nd workspace, and tiling is unaffected by the move.

### Displays

- [x] Windows move across displays in adjustment mode with direction keys
- [x] Windows move across displays with the mouse
- [ ] Changing the primary display moves the top bar. Window heights adjust on all monitors for the new position.
      - All windows moved over to new primary display
- [x] Unplug a display - windows from the display retile on a new workspace on the remaining display
- [ ] Plug in a display - windows and workspaces don't change
      - display had been primary display before unplugging it in previous test. Upon plugging in display again external display became primary monitor and all windows moved over there
- [ ] NOTE: Add vertical monitor layout test

### Launcher

- [x] All windows on all workspaces appear on launch
- [x] Choosing an app on another workspace moves workspaces and focus to that app
- [x] Launching an application works
- [x] Typing text and then removing it will re-show those windows
- [x] Search works for applications and windows
- [x] The overlay hint correctly highlights the selected window
- [x] t: executes a command in a terminal
- [x] : executes a command in sh
- [x] = calculates an equation

### Window Titles

- [x] Disabling window titles in Firefox works (Check debian and flatpak packages)

## With Tiling Disabled

### Tiling

- [x] Super direction keys changes focus in the correct direction
- [x] Windows can be moved with the keyboard
- [x] Windows can be moved with the mouse
- [x] Windows swap with the keyboard (test with different size windows)
- [x] Windows can be resized with the keyboard (Test resizing four windows above, below, right, and left to ensure shortcut consistency)
- [x] Windows can be resized with the mouse

### Displays

- [x] Windows move across displays in adjustment mode with directions keys
- [x] Windows move across displays with the mouse

**Comment by brs17** [comment]
All of the following patches introduced today seem to behave as expected:
- suppport loading icons from absolute paths
- nodisplay and hidden app entries...
- return icon widget from icon method
- fix launcher duplicates

Other test results are below. Note feedback in bold is feedback from this test today. I am now going to investigate with the same setup if these are regressions.

## With tiling enabled

### Tiling

- [x] Super direction keys changes focus in the correct direction
- [x] Windows moved with the keyboard tile into place
- [x] Windows moved with the mouse tile into place
- [x] Windows swap with the keyboard (test with different size windows)
- [x] Windows can be resized with the keyboard (Test resizing four windows above, below, right, and left to ensure shortcut consistency)
- [x] Windows can be resized with the mouse
- [x] Minimizing a window detaches it from the tree and re-tiles remaining windows
- [x] Unminimizing a window re-tiles it into its previous location
  - mostly works but has bug: https://github.com/pop-os/shell/issues/307
- [x] Maximizing removes the active hint and covers tiled windows
- [x] Umaximizing adds active hint and re-tiles into place
- [x] Full-screening removes the hint and full-screens on one display 
  - Not required, currently has a bug: https://github.com/pop-os/shell/issues/222
- [x] Unfull-screening adds active hint and re-tiles into place (Not required, currently has a bug)
- [x] Maximizing a YouTube video fills the screen and unmaximizing retiles the browser in place
  - fullscreen bug is present here too
- [x] VIM shortcuts work as direction keys
- [x] `Super` `O` changes window orientation
- [x] `Super` `G` floats and then re-tiles a window
- [x] `Super` `Q` Closes a window
- [x] `Super` `M` Maximizes and un-maximizes a window
- [x] Turn off auto-tiling. New windows launch floating.
- [x] Turn on auto-tiling. Windows automatically tile.
- [ ] Disabling and enabling auto-tiling correctly handles minimized, maximized, fullscreen, floating, and non-floating windows
- [ ] Float a window with `Super` `G`. It should be movable and resizeable in window management mode with keyboard keys
  - Not required, currently has a bug: https://github.com/pop-os/shell/issues/155

### Workspaces

- [ ] Windows can be moved to another workspace with the keyboard
  - **when moving window back to orignal workspace, windows from external display move back to primary display**
- [ ] Windows can be moved to another workspace with the mouse
- [ ] Windows can be moved to workspaces between existing workspaces
  - Works except when moving to a new workspace at the very beginning of the stack. 
- [x] Moving windows to another workspace re-tiled the previous and new workspace
- [x] Active hint is present on the new workspace and once the window is returned to its previous workspace
  - Broke, but commit is made that should fix it: https://github.com/pop-os/shell/pull/300/commits/255c01d11c6f4f327e93137b310201ee42351635
- [x] Floating windows move across workspaces
- [x] Remove windows from the 2nd worspace in a 3 workspace setup. The 3rd workspace becomes the 2nd workspace, and tiling is unaffected by the move.

### Displays

- [x] Windows move across displays in adjustment mode with direction keys
- [x] Windows move across displays with the mouse
- [ ] Changing the primary display moves the top bar. Window heights adjust on all monitors for the new position.
  - **All windows moved to new primary display**
- [ ] Unplug a display - windows from the display retile on a new workspace on the remaining display
- [ ] Plug in a display - windows and workspaces don't change
  - Still buggy: https://github.com/pop-os/shell/issues/171
  - **Unplugging, then plugging display back in (when system booted with display plugged in), creates odd moving windows (see video)**
- [ ] NOTE: Add vertical monitor layout test

### Launcher

- [x] All windows on all workspaces appear on launch
- [x] Choosing an app on another workspace moves workspaces and focus to that app
- [x] Launching an application works
- [x] Typing text and then removing it will re-show those windows
- [x] Search works for applications and windows
- [x] The overlay hint correctly highlights the selected window
- [x] t: executes a command in a terminal
- [x] : executes a command in sh
- [x] = calculates an equation

### Window Titles

- [x] Disabling window titles in Firefox works (Check debian and flatpak packages)
  - This may make firefox windows move around randomly when toggled, and you may have to restart firefox for it to work properly, but it should work besides that.

## With Tiling Disabled

### Tiling

- [x] Super direction keys changes focus in the correct direction
- [x] Windows can be moved with the keyboard
- [x] Windows can be moved with the mouse
- [ ] Windows swap with the keyboard (test with different size windows)
  - **windows change size to the size the window was in that location previously**
- [x] Windows can be resized with the keyboard (Test resizing four windows above, below, right, and left to ensure shortcut consistency)
- [x] Windows can be resized with the mouse
- [ ] Windows can be half-tiled left and right with `Ctrl``Super``left`/`right`
  - **does not work with vim keys**

### Displays

- [x] Windows move across displays in adjustment mode with directions keys
  - In floating mode, this works unless the windows have to cross the top bar. Bug: https://github.com/pop-os/shell/issues/184
- [x] Windows move across displays with the mouse

**Comment by brs17** [comment]
Update:

Went through the bolded comments and unbolded a few as they were not regressions. Added additional information where available.

**Comment by WatchMkr** [comment]
## Checklists

Tasks for a tester to verify when approving a patch. Use complex window layouts and at least two displays. Turn on active hint during testing.

## With tiling enabled

### Tiling

- [x] Super direction keys changes focus in the correct direction
- [x] Windows moved with the keyboard tile into place
- [x] Windows moved with the mouse tile into place
- [x] Windows swap with the keyboard (test with different size windows)
- [x] Windows can be resized with the keyboard (Shortcuts not consistent - NOT a regression)
- [x] Windows can be resized with the mouse
- [x] Minimizing a window detaches it from the tree and re-tiles remaining windows
- [ ] Unminimizing a window re-tiles it into its previous location **[REGRESSION]**
- [x] Maximizing removes the active hint and covers tiled windows
- [ ] Umaximizing adds active hint and re-tiles into place **[REGRESSION]**
- [x] Full-screening removes the hint and full-screens on one display 
- [x] Unfull-screening adds active hint and re-tiles into place (Not required, currently has a bug)
- [x] Maximizing a YouTube video fills the screen and unmaximizing retiles the browser in place
- [x] VIM shortcuts work as direction keys
- [ ] `Super` `O` changes window orientation **[REGRESSION]**
- [x] `Super` `G` floats and then re-tiles a window
- [x] `Super` `Q` Closes a window
- [x] `Super` `M` Maximizes and un-maximizes a window
- [x] Turn off auto-tiling. New windows launch floating.
- [x] Turn on auto-tiling. Windows automatically tile.
- [x] Disabling and enabling auto-tiling correctly handles minimized, maximized, fullscreen, floating, and non-floating windows - Looks good but the test needs a better description or perhaps separating out.
- [ ] Float a window with `Super` `G`. It should be movable and resizeable in window management mode with keyboard keys - Not required, currently has a bug: https://github.com/pop-os/shell/issues/155

### Workspaces

- [x] Windows can be moved to another workspace with the keyboard
- [x] Windows can be moved to another workspace with the mouse
- [x] Windows can be moved to workspaces between existing workspaces
  - Works except when moving to a new workspace at the very beginning of the stack. 
- [x] Moving windows to another workspace re-tiled the previous and new workspace
- [x] Active hint is present on the new workspace and once the window is returned to its previous workspace
- [x] Floating windows move across workspaces
- [x] Remove windows from the 2nd workspace in a 3 workspace setup. The 3rd workspace becomes the 2nd workspace, and tiling is unaffected by the move.

### Displays

- [x] Windows move across displays in adjustment mode with direction keys
- [x] Windows move across displays with the mouse
- [ ] Changing the primary display moves the top bar. Window heights adjust on all monitors for the new position. **CHECK IF THIS IS A REGRESSION**
- [x] Unplug a display - windows from the display retile on a new workspace on the remaining display
- [x] Plug in a display - windows and workspaces don't change

### Launcher

- [x] All windows on all workspaces appear on launch
- [x] Choosing an app on another workspace moves workspaces and focus to that app
- [x] Launching an application works
- [x] Typing text and then removing it will re-show those windows
- [x] Search works for applications and windows
- [x] The overlay hint correctly highlights the selected window
- [x] t: executes a command in a terminal
- [x] : executes a command in sh
- [x] = calculates an equation

### Window Titles

- [x] Disabling window titles in Firefox works (Check debian and flatpak packages)
  - This may make firefox windows move around randomly when toggled, and you may have to restart firefox for it to work properly, but it should work besides that.

## With Tiling Disabled

### Tiling

- [x] Super direction keys changes focus in the correct direction
- [ ] Windows can be moved with the keyboard **[REGRESSION]**
- [x] Windows can be moved with the mouse
- [x] Windows swap with the keyboard (test with different size windows)
- [ ] Windows can be resized with the keyboard **[REGRESSION]**
- [x] Windows can be resized with the mouse
- [x] Windows can be half-tiled left and right with `Ctrl``Super``left`/`right`

### Displays

- [ ] Windows move across displays in adjustment mode with directions keys **[REGRESSION]**
- [x] Windows move across displays with the mouse

**Comment by brs17** [comment]
https://github.com/pop-os/shell/pull/300/commits/30558c94507043a0feeec77c4ed269eb56112453 does not seem to be working, when unminimizing a window the window will be tiled next to the previously active window.

**Comment by WatchMkr** [comment]
## Checklists

Tasks for a tester to verify when approving a patch. Use complex window layouts and at least two displays. Turn on active hint during testing.

## With tiling enabled

### Tiling

- [x] Super direction keys changes focus in the correct direction
- [x] Windows moved with the keyboard tile into place
- [x] Windows moved with the mouse tile into place
- [x] Windows swap with the keyboard (test with different size windows)
- [x] Windows can be resized with the keyboard (Shortcuts not consistent - NOT a regression)
- [x] Windows can be resized with the mouse
- [x] Minimizing a window detaches it from the tree and re-tiles remaining windows
- [x] Unminimizing a window re-tiles it into its previous location (Restoring previous location could use improvement)
- [x] Maximizing removes the active hint and covers tiled windows
- [x] Umaximizing adds active hint and re-tiles into place
- [x] Full-screening removes the hint and full-screens on one display 
- [x] Unfull-screening adds active hint and re-tiles into place
- [x] Maximizing a YouTube video fills the screen and unmaximizing retiles the browser in place
- [x] VIM shortcuts work as direction keys
- [x] `Super` `O` changes window orientation
- [x] `Super` `G` floats and then re-tiles a window
- [x] `Super` `Q` Closes a window
- [x] `Super` `M` Maximizes and un-maximizes a window
- [x] Turn off auto-tiling. New windows launch floating.
- [x] Turn on auto-tiling. Windows automatically tile.
- [x] Disabling and enabling auto-tiling correctly handles minimized, maximized, fullscreen, floating, and non-floating windows - Looks good but the test needs a better description or perhaps separating out.
- [x] Float a window with `Super` `G`. It should be movable and resizeable in window management mode with keyboard keys

### Workspaces

- [x] Windows can be moved to another workspace with the keyboard
- [x] Windows can be moved to another workspace with the mouse
- [x] Windows can be moved to workspaces between existing workspaces
  - Works except when moving to a new workspace at the very beginning of the stack. 
- [x] Moving windows to another workspace re-tiled the previous and new workspace
- [x] Active hint is present on the new workspace and once the window is returned to its previous workspace
- [x] Floating windows move across workspaces
- [x] Remove windows from the 2nd workspace in a 3 workspace setup. The 3rd workspace becomes the 2nd workspace, and tiling is unaffected by the move.

### Displays

- [x] Windows move across displays in adjustment mode with direction keys
- [x] Windows move across displays with the mouse
- [x] Changing the primary display moves the top bar. Window heights adjust on all monitors for the new position.
- [x] Unplug a display - windows from the display retile on a new workspace on the remaining display
- [x] Plug in a display - windows and workspaces don't change

### Launcher

- [x] All windows on all workspaces appear on launch
- [x] Choosing an app on another workspace moves workspaces and focus to that app
- [x] Launching an application works
- [x] Typing text and then removing it will re-show those windows
- [x] Search works for applications and windows
- [ ] Search works for GNOME settings panels **[REGRESSION]**
- [x] The overlay hint correctly highlights the selected window
- [x] t: executes a command in a terminal
- [x] : executes a command in sh
- [x] = calculates an equation

### Window Titles

- [x] Disabling window titles in Firefox works (Check debian and flatpak packages)

## With Tiling Disabled

### Tiling

- [x] Super direction keys changes focus in the correct direction
- [x] Windows can be moved with the keyboard
- [x] Windows can be moved with the mouse
- [x] Windows swap with the keyboard (test with different size windows)
- [x] Windows can be resized with the keyboard
- [x] Windows can be resized with the mouse
- [] Windows can be half-tiled left and right with `Ctrl``Super``left`/`right`**[Left works | Right doesn't]

### Displays

- [x] Windows move across displays in adjustment mode with directions keys
- [x] Windows move across displays with the mouse

**Comment by mmstick** [comment]
This requires https://github.com/pop-os/gnome-control-center/pull/96

**Comment by WatchMkr** [comment]
## Checklists

Tasks for a tester to verify when approving a patch. Use complex window layouts and at least two displays. Turn on active hint during testing.

## With tiling enabled

### Tiling

- [x] Super direction keys changes focus in the correct direction
- [x] Windows moved with the keyboard tile into place
- [x] Windows moved with the mouse tile into place
- [x] Windows swap with the keyboard (test with different size windows)
- [x] Windows can be resized with the keyboard (Shortcuts not consistent - NOT a regression)
- [x] Windows can be resized with the mouse
- [x] Minimizing a window detaches it from the tree and re-tiles remaining windows
- [x] Unminimizing a window re-tiles it into its previous location (Restoring previous location could use improvement)
- [x] Maximizing removes the active hint and covers tiled windows
- [x] Umaximizing adds active hint and re-tiles into place
- [x] Full-screening removes the hint and full-screens on one display 
- [x] Unfull-screening adds active hint and re-tiles into place
- [x] Maximizing a YouTube video fills the screen and unmaximizing retiles the browser in place
- [x] VIM shortcuts work as direction keys
- [x] `Super` `O` changes window orientation
- [x] `Super` `G` floats and then re-tiles a window
- [x] `Super` `Q` Closes a window
- [x] `Super` `M` Maximizes and un-maximizes a window
- [x] Turn off auto-tiling. New windows launch floating.
- [x] Turn on auto-tiling. Windows automatically tile.
- [x] Disabling and enabling auto-tiling correctly handles minimized, maximized, fullscreen, floating, and non-floating windows - Looks good but the test needs a better description or perhaps separating out.
- [x] Float a window with `Super` `G`. It should be movable and resizeable in window management mode with keyboard keys

### Workspaces

- [x] Windows can be moved to another workspace with the keyboard
- [x] Windows can be moved to another workspace with the mouse
- [x] Windows can be moved to workspaces between existing workspaces
  - Works except when moving to a new workspace at the very beginning of the stack. 
- [x] Moving windows to another workspace re-tiled the previous and new workspace
- [x] Active hint is present on the new workspace and once the window is returned to its previous workspace
- [x] Floating windows move across workspaces
- [x] Remove windows from the 2nd workspace in a 3 workspace setup. The 3rd workspace becomes the 2nd workspace, and tiling is unaffected by the move.

### Displays

- [x] Windows move across displays in adjustment mode with direction keys
- [x] Windows move across displays with the mouse
- [x] Changing the primary display moves the top bar. Window heights adjust on all monitors for the new position.
- [x] Unplug a display - windows from the display retile on a new workspace on the remaining display
- [x] Plug in a display - windows and workspaces don't change

### Launcher

- [x] All windows on all workspaces appear on launch
- [x] Choosing an app on another workspace moves workspaces and focus to that app
- [x] Launching an application works
- [x] Typing text and then removing it will re-show those windows
- [x] Search works for applications and windows
- [x] Search works for GNOME settings panels
- [x] The overlay hint correctly highlights the selected window
- [x] t: executes a command in a terminal
- [x] : executes a command in sh
- [x] = calculates an equation

### Window Titles

- [x] Disabling window titles in Firefox works (Check debian and flatpak packages)

## With Tiling Disabled

### Tiling

- [x] Super direction keys changes focus in the correct direction
- [x] Windows can be moved with the keyboard
- [x] Windows can be moved with the mouse
- [x] Windows swap with the keyboard (test with different size windows)
- [x] Windows can be resized with the keyboard
- [x] Windows can be resized with the mouse

### Displays

- [x] Windows move across displays in adjustment mode with directions keys
- [x] Windows move across displays with the mouse

**Pull request merged**